src/crypto/hsm/nxp/PersistentStorageOperationalKeystoreHSM.h

```diff
@@ -0,0 +1,111 @@
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#pragma once
+
+#include "CHIPCryptoPALHsm_SE05X_utils.h"
+#include <crypto/OperationalKeystore.h>
+#include <lib/core/CHIPError.h>
+#include <lib/core/CHIPPersistentStorageDelegate.h>
+#include <lib/core/DataModelTypes.h>
+#include <lib/support/CHIPMem.h>
+#include <lib/support/CodeUtils.h>
+
+#if ENABLE_HSM_GENERATE_EC_KEY
+
+namespace chip {
+
+class PersistentStorageOperationalKeystoreHSM : public Crypto::OperationalKeystore
+{
+public:
+    PersistentStorageOperationalKeystoreHSM() = default;
+    virtual ~PersistentStorageOperationalKeystoreHSM() { Finish(); }
+
+    // Non-copyable
+    PersistentStorageOperationalKeystoreHSM(PersistentStorageOperationalKeystoreHSM const &) = delete;
+    void operator=(PersistentStorageOperationalKeystoreHSM const &) = delete;
+
+    /**
+     * @brief Initialize the Operational Keystore to map to a given storage delegate.
+     *
+     * @param storage Pointer to persistent storage delegate to use. Must outlive this instance.
+     * @retval CHIP_NO_ERROR on success
+     * @retval CHIP_ERROR_INCORRECT_STATE if already initialized
+     */
+    CHIP_ERROR Init(PersistentStorageDelegate * storage)
+    {
+        VerifyOrReturnError(mStorage == nullptr, CHIP_ERROR_INCORRECT_STATE);
+        mPendingFabricIndex       = kUndefinedFabricIndex;
+        mIsExternallyOwnedKeypair = false;
+        mStorage                  = storage;
+        mPendingKeypair           = nullptr;
+        mIsPendingKeypairActive   = false;
+        return CHIP_NO_ERROR;
+    }
+
+    /**
+     * @brief Finalize the keystore, so that subsequent operations fail
+     */
+    void Finish()
+    {
+        VerifyOrReturn(mStorage != nullptr);
+
+        ResetPendingKey();
+        mStorage = nullptr;
+    }
+
+    bool HasPendingOpKeypair() const override { return (mPendingKeypair != nullptr); }
+
+    bool HasOpKeypairForFabric(FabricIndex fabricIndex) const override;
+    CHIP_ERROR NewOpKeypairForFabric(FabricIndex fabricIndex, MutableByteSpan & outCertificateSigningRequest) override;
+    CHIP_ERROR ActivateOpKeypairForFabric(FabricIndex fabricIndex, const Crypto::P256PublicKey & nocPublicKey) override;
+    CHIP_ERROR CommitOpKeypairForFabric(FabricIndex fabricIndex) override;
+    CHIP_ERROR RemoveOpKeypairForFabric(FabricIndex fabricIndex) override;
+    void RevertPendingKeypair() override;
+    CHIP_ERROR SignWithOpKeypair(FabricIndex fabricIndex, const ByteSpan & message,
+                                 Crypto::P256ECDSASignature & outSignature) const override;
+    Crypto::P256Keypair * AllocateEphemeralKeypairForCASE() override;
+    void ReleaseEphemeralKeypair(Crypto::P256Keypair * keypair) override;
+
+protected:
+    void ResetPendingKey()
```

Comment:
Doesn't this need to clear the pending state on the slot that's pending?

Reply:
clearing the slot also now.